### PR TITLE
Initial support for multiple concurrent SQLAlchemy engines

### PR DESCRIPTION
**Pull request by jessemyers**
SQLAlchemy's `sessionmaker` supports a custom session type that, in turn, can override
`get_bind()` to control which engine it uses. Since the SQLAlchemy engine encompasses
the underlying Postgres connection and credentials, this approach allows us to use
different connections based on runtime context.

We can combine this notion with the `microcosm` "scoped factory" feature to allow multiple
configurations of the same entity to be managed simultaneously, with shared defines.

There are two caveats:
 1. The strategy used here to implement `get_bind()` isn't generic enough to meet our
    current anticipated use cases (e.g. connection-switching based on user auth)
 2. The scoped factory feature uses object proxies to manage object graph references;
    we have to be careful to unwrap these references to avoid breaking SQLAlchemy internals
    that performance object identity comparisons.

### CODE REVIEW

File: microcosm_postgres/factories/sessionmaker.py [hunk 1]
```diff
@@ -1,9 +1,47 @@
...
+        ScopedFactory.infect(graph, "postgres")
+
+    class RoutingSession(Session):
+        """
+        Route session bind to an appropriate engine.
+
+        See: http://docs.sqlalchemy.org/en/latest/orm/persistence_techniques.html#partitioning-strategies
+
+        """
+        def get_bind(self, mapper=None, clause=None):
```
Comment (jessemyers) [inline microcosm_postgres/factories/sessionmaker.py]: My implementation of `get_bind` implements a strategy based on the model class being used. This is not actually what we want, but it does demonstrate the ability.

File: microcosm_postgres/factories/sessionmaker.py [hunk 2]
```diff
@@ -1,9 +1,47 @@
...
+from microcosm.scoping.factories import ScopedFactory
+from microcosm.scoping.proxies import ScopedProxy
+from sqlalchemy.orm import Session
 from sqlalchemy.orm import sessionmaker
 
 
+ENGINE = "__engine__"
+
+
+def unwrap(bind):
```
Comment (jessemyers) [inline microcosm_postgres/factories/sessionmaker.py]: Note that we have to `unwrap` proxies.

Note further that even though the `ScopedFactory` usage on L27 is conditional, the way `ScopedFact.infect` is implemented is... well... infectious and doesn't "undo" when we reset unit test context. As a result, we need the `unwrap` ability even in unrelated tests. This was annoying to debug.

File: microcosm_postgres/tests/fixtures/employee.py [hunk 3]
```diff
@@ -16,13 +16,25 @@ class Employee(EntityMixin, Model):
...
         yield (self.company_id, self.id)
 
 
+class EmployeeData(EntityMixin, Model):
+    """
+    An employee data record containing sensitive data (accessed with a different engine).
+
+    """
+    __tablename__ = "employee_data"
+    __engine__ = "secret"
```
Comment (jessemyers) [inline microcosm_postgres/tests/fixtures/employee.py]: Notice me.

File: microcosm_postgres/factories/sessionmaker.py [hunk 4]
```diff
@@ -1,9 +1,47 @@
...
+    return bind.__component__ if isinstance(bind, ScopedProxy) else bind
+
+
+@defaults(
+    # enable multiple engines (binds) via ScopedFactory
+    multi=typed(boolean, default_value=True)
```
Comment (GabeLambda) [inline microcosm_postgres/factories/sessionmaker.py]: Defaults to multi?

Comment (jessemyers) [reply microcosm_postgres/factories/sessionmaker.py]: Good catch. Left over from local testing.

**Pull request closed without merging**